core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +668,934 @@ impl CheckedAdd for Fixed64 {
+					let mut t = 0;
+
+					if let Some(v) = u.checked_sub(v) {
+						if let Some(v2) = v.checked_sub(k) {
+							t = v2 % B;
+							k = 0;
+						} else {
+							needs_borrow = true;
+						}
+					} else {
+						needs_borrow = true;
+					}
+					if needs_borrow {
+						t = u + B - v - k;
+						k = 1;
+					}
+					t
+				};
+				// PROOF: t either comes from `v2 % B`, or from `u + B - v - k`. The former is
+				// trivial. The latter will not overflow this branch will only happen if the sum of
+				// `u - v - k` part has been negative, hence `u + B - v - k < b`.
+				w.set(j, s as Single);
+			}
+
+			if k.is_zero() {
+				Ok(w)
+			} else {
+				Err(w)
+			}
+		}
+
+		/// Multiplies n-limb number `self` with m-limb number `other`.
+		///
+		/// The resulting number will always have `n + m` limbs.
+		///
+		/// This function does not strip the output and returns the original allocated `n + m`
+		/// limbs. The caller may strip the output if desired.
+		///
+		/// Taken from "The Art of Computer Programming" by D.E. Knuth, vol 2, chapter 4.
+		pub fn mul(self, other: &Self) -> Self {
+			let n = self.len();
+			let m = other.len();
+			let mut w = Self::with_capacity(m + n);
+
+			for j in 0..n {
+				if self.get(j) == 0 {
+					// Note: `with_capacity` allocates with 0. Explicitly set j + m to zero if
+					// otherwise.
+					continue;
+				}
+
+				let mut k = 0;
+				for i in 0..m {
+					// PROOF: (B−1) × (B−1) + (B−1) + (B−1) = B^2 −1 < B^2. addition is safe.
+					let t =
+						mul_single(self.get(j), other.get(i))
+						+ Double::from(w.get(i + j))
+						+ Double::from(k);
+					w.set(i + j, (t % B) as Single);
+					// PROOF: (B^2 - 1) / B < B. conversion is safe.
+					k = (t / B) as Single;
+				}
+				w.set(j + m, k);
+			}
+			w
+		}
+
+		/// Divides `self` by a single limb `other`. This can be used in cases where the original
+		/// division cannot work due to the divisor (`other`) being just one limb.
+		///
+		/// Invariant: `other` cannot be zero.
+		pub fn div_unit(self, mut other: Single) -> Self {
+			other = other.max(1);
+			let n = self.len();
+			let mut out = Self::with_capacity(n);
+			let mut r: Single = 0;
+			// PROOF: (B-1) * B + (B-1) still fits in double
+			let with_r = |x: Double, r: Single| { r as Double * B + x };
+			for d in (0..=n-1).rev() {
+				let (q, rr) = div_single(with_r(self.get(d).into(), r), other) ;
+				out.set(d, q as Single);
+				r = rr;
+			}
+			out
+		}
+
+		/// Divides an `n + m` limb self by a `n` limb `other`. The result is a `m + 1` limb
+		/// quotient and a `n` limb remainder, if enabled by passing `true` in `rem` argument, both
+		/// in the form of an option's `Ok`.
+		///
+		/// - requires `other` to be stripped and have no leading zeros.
+		/// - requires `self` to be stripped and have no leading zeros.
+		/// - requires `other` to have at least two limbs.
+		/// - requires `self` to have a greater or equal length compared to `other`.
```

Comment:
I feel those constrains are quite strong maybe this could be reduced.

Actually I wonder if we want user to use this BigUInt directly maybe the most exposed API should be with less contrained as require and returning stripped output always ?

like we would rename this div_stripped_multiple_limbs and expose a div which only constrain is that other is different from 0 otherwise it returns 0